sql/opt_trace_ddl_info.cc

```diff
@@ -0,0 +1,206 @@
+  The record has full table_name,
+  and ddl of the table/view.
+*/
+static void dump_table_definition(THD *thd, DDL_Info_Record *rec)
+{
+  Json_writer_object ddl_wrapper(thd);
+  ddl_wrapper.add("name", rec->name);
+  size_t non_esc_stmt_len= strlen(rec->stmt);
+  size_t escape_stmt_len= 4 * non_esc_stmt_len;
+  char *escaped_stmt=
+      (char *) my_malloc(key_memory_trace_ddl_info, escape_stmt_len + 1, 0);
+  int act_escape_stmt_len=
+      json_escape_string(rec->stmt, rec->stmt + non_esc_stmt_len, escaped_stmt,
+                         escaped_stmt + escape_stmt_len);
+  escaped_stmt[act_escape_stmt_len]= 0;
+  ddl_wrapper.add("ddl", escaped_stmt);
+  my_free(escaped_stmt);
+}
+
+static void create_view_def(THD *thd, TABLE_LIST *table, String *name,
+                            String *stmt, String *view_def)
+{
+  view_def->append(STRING_WITH_LEN("CREATE "));
+  view_store_options(thd, table, view_def);
+  view_def->append(STRING_WITH_LEN("VIEW "));
+  view_def->append(*name);
+  view_def->append(STRING_WITH_LEN(" AS "));
+  view_def->append(*stmt);
+}
+
+/*
+  Stores the ddls of the tables, and views that are used
+  in either SELECT, INSERT, DELETE, and UPDATE queries,
+  into the optimizer trace.
+  Global query_tables are read in reverse order from the thd->lex,
+  and a record with table_name, and ddl of the table are created.
+  Hash is used to store the records, where in no duplicates
+  are stored. dbName_plus_tableName is used as a key to discard any
+  duplicates. If a new record that is created is not in the hash,
+  then that is dumped into the trace.
+*/
+void store_table_definitions_in_trace(THD *thd)
+{
+  LEX *lex= thd->lex;
+  if (thd->variables.optimizer_trace &&
+      thd->variables.store_ddls_in_optimizer_trace &&
+      (lex->sql_command == SQLCOM_SELECT ||
+       lex->sql_command == SQLCOM_INSERT ||
+       lex->sql_command == SQLCOM_INSERT_SELECT ||
+       lex->sql_command == SQLCOM_DELETE ||
+       lex->sql_command == SQLCOM_UPDATE ||
+       lex->sql_command == SQLCOM_DELETE_MULTI ||
+       lex->sql_command == SQLCOM_UPDATE_MULTI))
+  {
+    Json_writer_object ddls_wrapper(thd);
+    Json_writer_array ddl_list(thd, "list_ddls");
+    HASH hash;
+    List<TABLE_LIST> tables_list;
+
+    for (TABLE_LIST *tbl= thd->lex->query_tables; tbl; tbl= tbl->next_global)
+      tables_list.push_front(tbl);
+
+    if (tables_list.is_empty() || tables_list.elements == 0)
+      return;
+
+    List_iterator li(tables_list);
+    my_hash_init(key_memory_trace_ddl_info, &hash, system_charset_info, 16, 0,
+                 0, get_rec_key, free_rec, HASH_UNIQUE);
+    while (TABLE_LIST *tbl= li++)
+    {
+      if (tbl->is_view() ||
+          (!is_json_schema_nonuser_table(tbl) ||
+           (tbl->table && tbl->table->s &&
+            tbl->table->s->tmp_table == TRANSACTIONAL_TMP_TABLE)))
+      {
+        size_t name_len=
+            tbl->get_table_name().length + tbl->get_db_name().length + 1;
```

Comment:
So, this allocates memory for db_name, the '.' and the table_name. Where is the space for \0 at the end? 
Why not use String object here?